Src/FluentAssertions/Types/MethodBaseAssertions.cs

```diff
@@ -41,20 +44,23 @@ public AndConstraint<TAssertions> HaveAccessModifier(
     {
         Guard.ThrowIfArgumentIsOutOfRange(accessModifier);
 
-        bool success = Execute.Assertion
+        assertionChain
             .BecauseOf(because, becauseArgs)
             .ForCondition(Subject is not null)
-            .FailWith($"Expected method to be {accessModifier}{{reason}}, but {{context:member}} is <null>.");
+            .FailWith($"Expected method to be {accessModifier}{{reason}}, but {{context:method}} is <null>.");
 
-        if (success)
+        if (assertionChain.Succeeded)
         {
             CSharpAccessModifier subjectAccessModifier = Subject.GetCSharpAccessModifier();
 
-            Execute.Assertion
+            var subject = assertionChain.HasOverriddenCallerIdentifier
+                ? assertionChain.CallerIdentifier
+                : "method " + Subject.ToFormattedString();
```

Comment:
You're right. I missed that and will need to lazy-evaluate that everywhere.

Reply:
🤔Trying to come up with a new idiom for evaluating elements in the `FailWith` statement in a deferred way.

Reply:
Turns out that we already solved this problem. 

```
            assertionChain
                .ForCondition(!Subject!.CanRead)
                .BecauseOf(because, becauseArgs)
                .FailWith(() =>
                {
                    var subjectDescription = assertionChain.HasOverriddenCallerIdentifier
                        ? assertionChain.CallerIdentifier
                        : "property " + Subject.ToFormattedString();

                    return new FailReason($"Did not expect {subjectDescription} to have a getter{{reason}}.");
                });
```